.github/scripts/extract_git_ref.py

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python3
+
+import os
+
+def get_branch():
+    gh_event = os.environ["GITHUB_EVENT_NAME"]
+    gh_ref = os.environ["GITHUB_REF"]
+    gh_head_ref = os.environ.get("GITHUB_HEAD_REF")  # only exists in PRs
+
+    if gh_event == "pull_request":
+        if gh_head_ref is not None:
+            msg = f"BRANCH={gh_head_ref}"
+        else:
+            raise ValueError("GITHUB_HEAD_REF isn't set")
+    elif gh_event == "push":
+        # Handle differences between branches/tags
+        branch = "refs/heads/"
+        tag = "refs/tags/"
+        if gh_ref.startswith(branch):
+            msg = f"BRANCH={gh_ref[len(branch):]}"
+        elif gh_ref.startswith(tag):
+            msg = f"BRANCH={gh_ref[len(tag):]}"
+        else:
+            raise ValueError("Unanticipated git ref from a GitHub push event")
+    elif gh_event == "workflow_dispatch":
+        # Workflow Dispatch should only be able to trigger from branches
+        branch = "refs/heads/"
+        if gh_ref.startswith(branch):
+            msg = f"BRANCH={gh_ref[len(branch):]}"
+        else:
+            raise ValueError("Unanticipated git ref from a GitHub workflow_dispatch event")
+    else:
+        raise ValueError("Unanticipated GitHub Event")
+    return msg
+
+
+if __name__ == '__main__':
```

Comment:
Good catch. I thought I had run it through Black, but my editor wasn't picking up the environment in `services/api` and silently failed when it tried to call `black`.

I agree on getting these scripts set up with linting & etc... now that we're using Python in CI. I'm leaning towards adding a new CI job that just hand installs black/flake8/mypy/pytest, etc... and checks things in `.github/scripts/*.py`. How does that sound? The [GitHub runners are on Python 3.8.10](https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md#language-and-runtime) for now so I'm not sure we'd want to reuse the `services/api` environment even though that'd be a lot easier.

Reply:
I don’t think I like the idea of having a dev dependency like Black defined only in CI. I think maybe taking another look at something like [pre-commit](https://pre-commit.com/) for the entire repo, instead of having trying to define linters/formatters per-service.

Maybe we should poke around for guidance on dev dependencies like this in heterogeneous monorepos… I’ve seen a few monorepos recently, but they’re 100% Node services, so it makes sense to have something like `package.json` in the root of the repo. Maybe we can standardize on something like Python for dev tools, though, since we’re headed down that road with this CI script anyway.

Reply:
That's fair about not wanting the dev dependency to be solely captured in CI. I did briefly consider setting up a Poetry project or requirements.txt in `.github/scripts` so the dependencies were captured. I'll push what I currently have in case it's helpful to this discussion and to resolve a few of these other comments.

I think looking around for guidance on dev dependencies in monorepos could be worthwhile. I'm not sure about standardizing on one language for dev & build tooling - it seems like there would be less friction working with the language tooling natively vs trying to call Node stuff from Python. Pre-commit might be able to help us out with the git hooks though. I have also used [Bazel](https://bazel.build/) to stitch multiple languages together. It was nice compared to native C++ tooling but gets complicated compared to NPM/Poetry and we may need to write custom plugins.